libs/nio/src/main/java/org/elasticsearch/nio/SocketChannelContext.java

```diff
@@ -234,49 +234,107 @@ protected boolean closeNow() {
+    // by reusing a thread-local direct buffer (provided by the NioSelector).
+    //
+    // Each network event loop is given a 64kb DirectByteBuffer. When we read we use this buffer and copy the
+    // data after the read. When we go to write, we copy the data to the direct memory before calling write.
+    // The choice of 64KB is rather arbitrary. We can explore different sizes in the future. However, any
+    // data that is copied to the buffer for a write, but not successfully flushed immediately, must be
+    // copied again on the next call.
+
     protected int readFromChannel(ByteBuffer buffer) throws IOException {
+        ByteBuffer ioBuffer = getSelector().getIoBuffer();
+        ioBuffer.limit(Math.min(buffer.remaining(), ioBuffer.limit()));
+        int bytesRead;
         try {
-            int bytesRead = rawChannel.read(buffer);
-            if (bytesRead < 0) {
-                closeNow = true;
-                bytesRead = 0;
-            }
-            return bytesRead;
+            bytesRead = rawChannel.read(ioBuffer);
         } catch (IOException e) {
             closeNow = true;
             throw e;
         }
+        if (bytesRead < 0) {
+            closeNow = true;
+            return 0;
+        } else {
+            ioBuffer.flip();
+            buffer.put(ioBuffer);
+            return bytesRead;
+        }
     }
 
-    protected int readFromChannel(ByteBuffer[] buffers) throws IOException {
+    protected int readFromChannel(InboundChannelBuffer channelBuffer) throws IOException {
+        ByteBuffer ioBuffer = getSelector().getIoBuffer();
+        int bytesRead;
         try {
-            int bytesRead = (int) rawChannel.read(buffers);
-            if (bytesRead < 0) {
-                closeNow = true;
-                bytesRead = 0;
-            }
-            return bytesRead;
+            bytesRead = rawChannel.read(ioBuffer);
         } catch (IOException e) {
             closeNow = true;
             throw e;
         }
+        if (bytesRead < 0) {
+            closeNow = true;
+            return 0;
+        } else {
+            ioBuffer.flip();
+            channelBuffer.ensureCapacity(channelBuffer.getIndex() + ioBuffer.remaining());
+            ByteBuffer[] buffers = channelBuffer.sliceBuffersFrom(channelBuffer.getIndex());
+            int j = 0;
+            while (j < buffers.length && ioBuffer.remaining() > 0) {
+                ByteBuffer buffer = buffers[j++];
+                copyBytes(ioBuffer, buffer);
+            }
+            channelBuffer.incrementIndex(bytesRead);
+            return bytesRead;
+        }
     }
 
     protected int flushToChannel(ByteBuffer buffer) throws IOException {
+        ByteBuffer ioBuffer = getSelector().getIoBuffer();
+        copyBytes(buffer, ioBuffer);
+        ioBuffer.flip();
         try {
-            return rawChannel.write(buffer);
+            return rawChannel.write(ioBuffer);
```

Comment:
Shouldn't we move the position on `buffer` back again by the number of writes not written if the write doesn't succeed in writing all of `ioBuffer`, i.e. when `ioBuffer.remaining() < rawChannel.write(ioBuffer)`?

Reply:
This is an issue. The passed buffer needs to reflect partial flushes.